.github/windowdesc-ini_key.py

```diff
@@ -0,0 +1,48 @@
+"""
+Script to scan the OpenTTD source-tree for ini_key issues in WindowDesc entries.
+"""
+
+import glob
+import os
+import re
+import sys
+
+def scan_source_files(path, errors, ini_keys):
+    for new_path in glob.glob(f"{path}/*.cpp"):
+        if os.path.isdir(new_path):
+            errors.append(scan_source_files(new_path, errors, ini_keys))
+            continue
+
+        with open(new_path) as fp:
+            output = fp.read()
+
+        for windowdesc in re.findall(r"^static WindowDesc ([a-zA-Z0-9_]*).*?, (?:\"(.*?)\")?.*?,(?:\s+(.*?),){6}", output, re.S|re.M):
+            name = windowdesc[0]
+            ini_key = windowdesc[1]
+            widgets = windowdesc[2]
+            if ini_key:
+                if ini_key in ini_keys:
+                    errors.append(f"{new_path}: {name} ini_key is a duplicate")
+                ini_keys.add(ini_key)
+            widget = re.findall("static const (?:struct )?NWidgetPart " + widgets + ".*?(?:WWT_(DEFSIZE|STICKY)BOX.*?)?;", output, re.S)[0]
+            if widget and not ini_key:
+                errors.append(f"{new_path}: {name} has WWT_DEFSIZEBOX/WWT_STICKYBOX without ini_key")
+            if ini_key and not widget:
+                errors.append(f"{new_path}: {name} has ini_key without WWT_DEFSIZEBOX/WWT_STICKYBOX")
+
+
+def main():
+    errors = []
+    ini_keys = set()
+
+    scan_source_files("src", errors, ini_keys)
```

Comment:
better to create `errors` inside the function, and return it. That is a lot cleaner Python-wise.

Don't worry about the "copies" you make for the recursive calls; it is really not worth mentioning in this context :)